File: tagger-gui/src/test/java/no/lau/vdvil/dirigent/DirigentTest.java
package no.lau.vdvil.dirigent;

import no.lau.vdvil.cache.VdvilCache;
import org.codehaus.httpcache4j.cache.VdvilHttpCache;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * A Simple structuring of the dirigent
 */
public class DirigentTest {
    Logger log =  LoggerFactory.getLogger(DirigentTest.class);

    @Test
    public void testSetupAndPlaying() throws InterruptedException {
        Dirigent dirigent = new Dirigent("http://kpro09.googlecode.com/svn/trunk/graph-gui-scala/src/main/resources/composition/javazone.dvl.composition.xml");
        dirigent.addDownloader(VdvilHttpCache.create());
        dirigent.addHandler(new VdvilHandler() {
            public void load(InputStream inputStream, String mimeType) {
                log.info("Loading");
            }
        });
        dirigent.changePlayBackSpeed(150F);
        dirigent.prepare();
        dirigent.play(32);
        Thread.sleep(5000);
        dirigent.stop();
    }
}

class Dirigent {
    Logger log =  LoggerFactory.getLogger(Dirigent.class);

    private List<VdvilCache> downloaders = new ArrayList<VdvilCache>(); ;
    private List<VdvilHandler> handlers = new ArrayList<VdvilHandler>();
    private Float playbackBpm;
    private String compositionUrl;

    Dirigent(String compositionUrl) {
        this.compositionUrl = compositionUrl;
    }

    public void prepare() {
        for (VdvilCache downloader : downloaders) {
            if(downloader.acceptsUrl(compositionUrl)) {
                log.info("Trying to download {} with ", compositionUrl, downloader);
                InputStream inputStream = downloader.fetchAsStream(compositionUrl);

                String mimeType = downloader.mimeType(compositionUrl);
                log.info("{} has Mime type {}", compositionUrl, mimeType);
                for (VdvilHandler handler : handlers) {
                    handler.load(inputStream, mimeType);
                }

            }
            else {
                log.info("Downloader {} would not try to download {}", downloader,  compositionUrl);
            }
        }

    }

    public void addDownloader(VdvilCache downloader) { downloaders.add(downloader); }

    public void addHandler(VdvilHandler handler) { handlers.add(handler); }

    public void changePlayBackSpeed(Float playbackBpm) { this.playbackBpm = playbackBpm; }

    public void play(int startAtCue) {

    }

    public void stop() {

    }
}

interface VdvilHandler {

    void load(InputStream inputStream, String mimeType);
}

File: play-gui-scala/src/test/scala/DirigentTest.scala
package no.lau.vdvil.dirigent

import java.net.URL
import org.junit.{Ignore, Test}
import org.slf4j.LoggerFactory
import org.codehaus.httpcache4j.cache.VdvilHttpCache
import no.lau.vdvil.handler.persistence.{CompositionXMLParser, SimpleFileCache}
import no.vdvil.renderer.image.cacheinfrastructure.ImageDescriptionXMLParser
import no.vdvil.renderer.audio.AudioXMLParser
import no.lau.vdvil.handler.{Composition, DownloadAndParseFacade}
import no.lau.vdvil.cache.testresources.TestMp3s

/**
 * A Simple structuring of the Dirigent
 * The Dirigent is a simple generic way of handling the domain of multimedia playback
 *
 */
@Deprecated //Has been severely outdated
class DirigentTest {
  var compositionUrl = TestMp3s.javaZoneComposition

  val player = new no.lau.vdvil.player.ScalaCompositionPlayer(None, VdvilHttpCache.create())
  val log = LoggerFactory.getLogger(classOf[DirigentTest])

  @Test
  @Ignore def testSetupAndPlaying {

    val dirigent = new Dirigent(ScalaHttpCache :: LocalFileCache :: Nil, new CompositionHandler :: Nil, player) {
      playbackBpm = 150F
      prepare(compositionUrl)
    }
    dirigent.play(32)
    Thread.sleep(5000)
    dirigent.stop
  }

  @Test
  def advancedPreparing {
    val dirigent = new Dirigent(ScalaHttpCache :: LocalFileCache :: Nil, new CompositionHandler :: Nil, player)
    //dirigent.preparingMagic(compositionUrl)
  }
}

class ImprovedDirigentTest {
  @Test
  def runThis {
    val compositionURL = TestMp3s.javaZoneComposition
    val downloadAndParseFacade = new DownloadAndParseFacade {
      addCache(VdvilHttpCache.create())
      addParser(new CompositionXMLParser(this))
      addParser(new ImageDescriptionXMLParser(this))
      addParser(new AudioXMLParser(this))
    }
    val composition:Composition = downloadAndParseFacade.parse("", compositionURL).asInstanceOf[Composition]
    println("My compositiion; " + composition)
    /*
    val player = new ScalaCompositionPlayer(Some(composition), VdvilHttpCache.create)
    player.play(0, 150F)
    Thread.sleep((16*60/150)*1000)
    player.stop
    */
  }
}
